core/src/multi_proof.rs

```diff
@@ -0,0 +1,548 @@
+    // to reconstruct all other nodes later.
+    external_siblings: Vec<Node>,
+}
+
+impl MultiProof {
+    /// Construct a MultiProof from a vector of *ordered* PathProof
+    pub fn from_path_proofs(path_proofs: Vec<PathProof>) -> Self {
+        // A multi-proof can be viewed by associating each terminal node
+        // with its first n uniquely related siblings from its path proof,
+        // followed by all necessary siblings that are not derivable from
+        // the previously mentioned siblings.
+        let mut sub_paths: Vec<SubPathProof> = vec![];
+        let mut external_siblings: Vec<Node> = vec![];
+
+        // The goal is to traverse the entire tree
+        // formed by all path proofs and only collect the siblings
+        // that cannot be reconstructed in the future
+        //
+        // The traversal does not occur on the siblings themselves,
+        // but on the ordered set of key_paths within PathProofs.
+        //
+        // For example, take two key_paths starting from the first bit.
+        // If two keys share the same bit at index i, they will have the same sibling at index i.
+        // If the bit at index i differs, their paths diverge, and no sibling is needed at that index
+        // because it can be reconstructed from siblings collected later in the other key.
+        // When a differing bit is found, each key_path needs separate scanning.
+        //
+        // When there is a single key_path, all siblings are necessary.
+        //
+        // When there are more than two key_paths, the same logic is applied,
+        // but if two key_paths have different bits at an index, the key_paths are split
+        // based on the value at that index.
+        //
+        // If all key_paths share a bit at an index, that sibling is required;
+        // if at least two key_paths differ, no sibling is needed, but the key_paths must be divided
+        // based on having bit 0 or 1 at that index.
+        //
+        // Iterate this algorithm on the bisection to determine the minimum necessary siblings.
+        //
+        // `external_siblings` whill follow this structure for each bicection:
+        // |commons siblings| ext siblings in the right bisection | ext siblings in the left bisection |
+
+        // it represent a slice of the vector of key_paths with the notion of
+        // path_index, at which bit in the key_path that slice is pointing at
+        struct TraverseInfo {
+            path_index: usize,
+            path_slice_lower: usize,
+            path_slice_upper: usize,
+            ext_sibling_index: usize,
+        }
+        let mut stack_traverse_info: Vec<TraverseInfo> = vec![];
+
+        // External siblings encountered during a subtree traversal
+        let mut ext_sibling_in_sub_tree: Vec<Node> = vec![];
+
+        // initially we're looking at all the key_paths
+        let mut curr_traverse_info = TraverseInfo {
+            path_index: 0,
```

Comment:
I find the name `path_index` misleading. It looks like it'd be the index in `path_proofs`, but it's actually the bit index in the key path where the bisection was performed.
I recommend changing this to `bit_index` or `path_bit_index`.